### ISSUE TITLE
Stats CSV page: curl: (18) transfer closed with outstanding read data remaining

### ISSUE THREAD

**Issue by ghost** [haproxy/haproxy#1478]
### Detailed Description of the Problem

This is continuation of:
https://discourse.haproxy.org/t/stats-csv-page-curl-18-transfer-closed-with-outstanding-read-data-remaining/6745

I am trying to read stats csv page and sometimes I have problem:

`curl: (18) transfer closed with outstanding read data remaining`

I am running haproxy as docker container (tag: 2.4.4), I noticed, that when I restart container, then problem is gone for some time.

I've captured tcpdump as requested. It's attached.
[tcpdump.txt](https://github.com/haproxy/haproxy/files/7667881/tcpdump.txt)

### Expected Behavior

I can get stats csv content.

### Steps to Reproduce the Behavior

no clear steps.

### What is your configuration?

```haproxy
global
    log 127.0.0.1  local2
    maxconn        8100

defaults
    log global
    maxconn 1600

listen stats
    bind 0.0.0.0:9991
    mode http
    stats enable
    http-request use-service prometheus-exporter if { path /metrics }

listen BACKEND_NAME
    bind 0.0.0.0:8098
    bind 0.0.0.0:9990
    mode    tcp
    balance first
    log-format "%Ci:%Cp %Bi:%Bp [%t] %ft %b/%s %Tw/%Tc/%Tt %B %ts %ac/%fc/%bc/%sc/%rc %sq/%bq"
    option tcplog
    option redispatch
    option tcp-check
    option log-health-checks
    option httpchk GET /v1/health/state/critical
    http-check expect rstring ^\[\]$
    http-check disable-on-404
    timeout check 5s
    timeout client 35m
    timeout connect 600
    timeout queue 300s
    timeout server 35m
    timeout server-fin 2m
    default-server maxconn 1 check port 8500 inter 2s fastinter 1s downinter 30s rise 1 fall 2
    server servername-10.89.0.10 10.89.0.10
    server servername-10.89.0.11 10.89.0.11
    server servername-10.89.0.12 10.89.0.12
    server servername-10.89.0.13 10.89.0.13
    server servername-10.89.0.14 10.89.0.14
    server servername-10.89.0.15 10.89.0.15
    server servername-10.89.0.16 10.89.0.16
    server servername-10.89.0.17 10.89.0.17
    server servername-10.89.0.18 10.89.0.18
    server servername-10.89.0.19 10.89.0.19
    server servername-10.89.0.20 10.89.0.20
    server servername-10.89.0.21 10.89.0.21
    server servername-10.89.0.22 10.89.0.22
    server servername-10.89.0.23 10.89.0.23
    server servername-10.89.0.24 10.89.0.24
    server servername-10.89.0.25 10.89.0.25
    server servername-10.89.0.26 10.89.0.26
    server servername-10.89.0.27 10.89.0.27
    server servername-10.89.0.28 10.89.0.28
    server servername-10.89.0.29 10.89.0.29
    server servername-10.89.0.30 10.89.0.30
    server servername-10.89.0.31 10.89.0.31
    server servername-10.89.0.32 10.89.0.32
    server servername-10.89.0.33 10.89.0.33
    server servername-10.89.0.34 10.89.0.34
    server servername-10.89.0.35 10.89.0.35
    server servername-10.89.0.36 10.89.0.36
    server servername-10.89.0.37 10.89.0.37
    server servername-10.89.0.38 10.89.0.38
    server servername-10.89.0.39 10.89.0.39
    server servername-10.89.0.40 10.89.0.40
    server servername-10.89.0.41 10.89.0.41
    server servername-10.89.0.42 10.89.0.42
    server servername-10.89.0.43 10.89.0.43
    server servername-10.89.0.44 10.89.0.44
    server servername-10.89.0.45 10.89.0.45
    server servername-10.89.0.46 10.89.0.46
    server servername-10.89.0.47 10.89.0.47
    server servername-10.89.0.48 10.89.0.48
    server servername-10.89.0.49 10.89.0.49
    server servername-10.89.0.50 10.89.0.50
    server servername-10.89.0.51 10.89.0.51
    server servername-10.89.0.52 10.89.0.52
    server servername-10.89.0.53 10.89.0.53
    server servername-10.89.0.54 10.89.0.54
    server servername-10.89.0.55 10.89.0.55
    server servername-10.89.0.56 10.89.0.56
    server servername-10.89.0.57 10.89.0.57
    server servername-10.89.0.58 10.89.0.58
    server servername-10.89.0.59 10.89.0.59
    server servername-10.89.0.60 10.89.0.60
    server servername-10.89.0.61 10.89.0.61
    server servername-10.89.0.62 10.89.0.62
    server servername-10.89.0.63 10.89.0.63
    server servername-10.89.0.64 10.89.0.64
    server servername-10.89.0.65 10.89.0.65
    server servername-10.89.0.66 10.89.0.66
    server servername-10.89.0.67 10.89.0.67
    server servername-10.89.0.68 10.89.0.68
    server servername-10.89.0.69 10.89.0.69
    server servername-10.89.0.70 10.89.0.70
    server servername-10.89.0.71 10.89.0.71
    server servername-10.89.0.72 10.89.0.72
    server servername-10.89.0.73 10.89.0.73
    server servername-10.89.0.74 10.89.0.74
    server servername-10.89.0.75 10.89.0.75
    server servername-10.89.0.76 10.89.0.76
    server servername-10.89.0.77 10.89.0.77
    server servername-10.89.0.78 10.89.0.78
    server servername-10.89.0.79 10.89.0.79
    server servername-10.89.0.80 10.89.0.80
    server servername-10.89.0.81 10.89.0.81
    server servername-10.89.0.82 10.89.0.82
    server servername-10.89.0.83 10.89.0.83
    server servername-10.89.0.84 10.89.0.84
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.4-acb1d0b 2021/09/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.4.html
Running on: Linux 3.10.0-1127.19.1.el7.x86_64 #1 SMP Tue Aug 25 17:23:54 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
Running on OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

**Comment by capflam** [comment]
Could you try the 2.4.9 to check if the issue is still there ? Thanks !

**Comment by ghost** [comment]
Issue is still visible with 2.4.9.

**Comment by capflam** [comment]
Thanks ! I'm unable to reproduce the bug with your config snippet. I guess it is a time dependent issue. Could you provide a network capture ? Do you know if the problem is reproducible with no traffic ? If so, it could be good to enable H1 traces.

**Comment by lukastribus** [comment]
The network capture should be a pcap file (tcpdump -s0 -w file.cap), not the stdout summary from tcpdump.

**Comment by ghost** [comment]
Issue appears after some time (this time it took ~2days to reproduce). I've attached pcap file (if something is still wrong, please let me know, I am not experienced with network analise)
[file.cap.txt](https://github.com/haproxy/haproxy/files/7692297/file.cap.txt)
.

**Comment by capflam** [comment]
Thanks ! It appears the response is chunked (as expected with the stats page), but the payload itself is not chunked. I don't know for now how it is possible. But it is a good start to investigate.

**Comment by capflam** [comment]
My bad, the response is chunked but closed before the last chunk. BTW, I'm investigating.

**Comment by wtarreau** [comment]
We've already had issues in the past where a last block of exactly the size of a buffer minus epsilon couldn't be sent, but the final shut was already queued. I wouldn't completely rule out something a bit similar, which could explain why it only appears there. I've met myself a few times some truncated stats pages only after some specific uptime (due to number having to be large enough to fill buffers), but we fixed that a while ago. It could also be possible that the issue is in the stats applet itself (e.g. a missing check on a chunk_appendf()) resulting in one block being silently dropped.

**Comment by Heimschmiede** [comment]
This issues started to happen in our infrastructure. (2.4.15). We've got two ubuntu focal machines with identical setup (one active node + one failover node). The problem just happens on the passive node. When haproxy starts up everything is fine. The issue starts to appear after some hours and only affects the csv export in my case. The json export works fine.
Are there any news regarding this issue?

**Comment by wtarreau** [comment]
Since then we've fixed an issue where an occasional error between haproxy and a server could sometimes result in the last bytes being truncated. But in your case it's not a server, it's the stats page. Or are you accessing the passive node's stats page via the active node maybe ?

**Comment by Heimschmiede** [comment]
@wtarreau  good idea but no. I directly access the stats page via the machines ip address.


Upgraded on 2022-03-23 to version from `2.4.14-1ppa1~focal`  to  `2.4.15-1ppa1~focal`
Problem started to occur on 2022-03-31 (-> not directly related)

```
listen stats
    bind *:9090
    mode http
    http-request set-header X-Forwarded-Proto http
    stats enable
    stats hide-version
    stats realm Haproxy\ Statistics
    stats uri /
```

The last chunk seems to be missing. The data about the stats itself is missing starting with column `uweight`. See the last lines of the csv exports from my active and from the passive node: [csv_export.csv](https://github.com/haproxy/haproxy/files/8415148/csv_export.csv)

~I will try to downgrade to 2.4.14~
EDIT: sadly the 2.4.14-1 vesino is not available on ppa:vbernat/haproxy-2.4

**Comment by capflam** [comment]
I guess it is not the full csv export. What it is the size of a full export ? Do you know if you've hit a timeout ? If yes, it should be visible in haproxy logs.

**Comment by capflam** [comment]
If the problem appears on the passive node, it means there is no traffic on it, right ? Thus it may be possible to enable H1 traces when the problem appears. It is pretty verbose but it could really help to understand why the response is truncated. You can use the following script or something similar, from which you'll need to capture the output:

```
$ (echo "trace h1 sink buf0; trace h1 level developer; trace h1 verbosity complete; trace h1 start now;show events buf0 -w -n" ; cat ) | socat - HA-SOCKET
```

Note that all H1 traffic will be dump, on client and server side. It may contain sensitive info.

To disable traces:
```
$ echo "trace 0" | socat - HA-SOCKET
```

**Comment by Heimschmiede** [comment]
> I guess it is not the full csv export. What it is the size of a full export ? Do you know if you've hit a timeout ? If yes, it should be visible in haproxy logs.

correct. It was just the ending of the csv response (last line). A normal response size is about 36kB - 45kB in size.

> If the problem appears on the passive node, it means there is no traffic on it, right ? Thus it may be possible to enable H1 traces when the problem appears. It is pretty verbose but it could really help to understand why the response is truncated.

Yes, on the passive node is not much traffic. I turned tracing on for a short while and performed one csv download using chrome. See the output attached (I anonymized domain and cookies). Search for `GET /;csv`
[trace-log.txt](https://github.com/haproxy/haproxy/files/8418194/trace-log.txt)

**Comment by capflam** [comment]
Thanks ! I'll take a look. You are using HTTP compression and it may be related. Or it may be just a way to hit the bug more quickly.

**Comment by Heimschmiede** [comment]
Yes, let me provide my `global` and `default` config.
Btw: a friended team observed the same issue and fixed it for now by downgrading to version `2.2`
```
global
    maxconn 2048
    log-send-hostname
    log 127.0.0.1 local0
    log 127.0.0.1 local1 notice
    user haproxy
    group haproxy
    daemon
    stats socket /var/run/haproxy.sock mode 600 level admin
    stats timeout 2m
    # ...
defaults
    mode http
    log global
    option httplog
    timeout connect 5000
    timeout check 5000
    timeout client 60000
    timeout server 60000
    option forwardfor
    option http-server-close
    option http-ignore-probes
    option httpchk GET /healthcheck
    compression algo gzip
    compression type text/html text/plain text/css application/javascript application/json

# frontends / backends
# ...

listen stats
    bind *:9090
    mode http
    http-request set-header X-Forwarded-Proto http
    stats enable
    stats hide-version
    stats realm Haproxy\ Statistics
    stats uri /
```

**Comment by capflam** [comment]
Thanks for the info. I'm still investigating

**Comment by capflam** [comment]
Ok I'm now able to reproduce the bug. Or at least something similar. Just to know, have you a disabled proxy (frontend/backend/listener) at the end of your configuration or a peer section or something special ? I'm must read more carefully the code to fully understand the issue.

**Comment by Heimschmiede** [comment]
> Ok I'm now able to reproduce the bug. Or at least something similar. Just to know, have you a disabled proxy (frontend/backend/listener) at the end of your configuration or a peer section or something special ? I'm must read more carefully the code to fully understand the issue.

no, all backends and frontends are active and ready to accept requests. The only thing is that the frontends bind a virtual ip which is not assigned to the passive node. In my setup I assign a virtual ip managed be keepalived. That could at least explain why the active node is not running into this issue.

**Comment by capflam** [comment]
Ok, so it may be another way to hit the issue. In my case, the payload is not truncated but the last empty chunk is missing. So I must understand how stats can be truncated to properly fix the bug.

**Comment by capflam** [comment]
It should be fixed in 2.6-DEV. If you are interested, you can try it on the 2.4 with the following patch. Otherwise, you may wait for the backport and the next release (not planned for now). I can confirm the 2.2 is not affected. Thanks for your help !

[0001-BUG-MEDIUM-stats-Be-sure-to-never-set-EOM-flag-on-an.patch.2.4.txt](https://github.com/haproxy/haproxy/files/8441420/0001-BUG-MEDIUM-stats-Be-sure-to-never-set-EOM-flag-on-an.patch.2.4.txt)
